src/packaging/kpatch/commit_to_overlay.sh

```diff
@@ -0,0 +1,459 @@
+    read user_input
+
+    case "${user_input}" in
+      "y"*|"Y"*)
+        echo -en "Proceeding with push.\n\n"
+        ;;
+      "n"*|"N"*)
+        exit
+        ;;
+      "m"*|"M"*)
+        echo "Retaining ${LINUX_SRC} and branch ${OVERLAY_BRANCH} for manual push later."
+        MANUAL_PUSH=0
+        ;;
+      *)
+        echo -en "${COLOR_RED}ERROR: Invalid response.\nPlease enter 'y' to proceed with push," \
+                 "'n' to delete the branch and exit,\nor 'm' to retain the branch for manual" \
+                 "push at a later time.${NO_COLOR}\n\n"
+        ;;
+    esac
+  done
+}
+
+# Cleanup artifacts
+_cleanup() {
+  cd $RUN_DIR
+  echo -en "\nCleaning up and exiting\n"
+
+  if [[ -d ${LINUX_SRC} ]] && [[ -d ${VDO_TREE} ]] && [[ ${DEBUG} != 0 ]]; then
+    if [[ ${MANUAL_PUSH} != 0 ]]; then
+      echo "Removing ${LINUX_SRC}"
+      rm -rf ${LINUX_SRC}
+    fi
+
+    echo "Removing ${VDO_TREE}"
+    rm -rf ${VDO_TREE}
+  fi
+
+  if [[ -e ${commit_file} ]] && [[ -f ${commit_file} ]]; then
+    rm -fv ${commit_file}
+  fi
+
+  unset -f CHANGE_LOG DEBUG GITDATE GITAUTHOR LINUX_SRC
+}
+
+###############################################################################
+# main()
+trap _cleanup EXIT
+test "$#" -lt 5 && print_usage
+
+process_args $@
+print_config
+checkout_source_branch
+$PROCESS_INPUT_FX
+
+echo # Intentional blank line
+
+# Store linux repo location as an absolute path in the current directory
+LINUX_SRC=${RUN_DIR}/$(mktemp -d vdo-kernel-XXXXXX)
+VDO_TREE=${RUN_DIR}/$(mktemp -d vdo-raw-XXXXXX)
+
+# Shallow clone the specified dm-linux repo branch into LINUX_SRC
+echo "Cloning kernel repo $(get_repo_name ${KERNEL_REPO_URL})/${KERNEL_BRANCH}"
+git clone --branch ${KERNEL_BRANCH} --depth 1 ${KERNEL_REPO_URL} ${LINUX_SRC}
+cd $LINUX_SRC
+git checkout -b ${OVERLAY_BRANCH} ${KERNEL_BRANCH}
+
+# Need to detect that our repo is x number of commits ahead of dm-vdo/vdo-devel
+# Clone the current repo as ${VDO_TREE}? Is VDO_TREE even needed? Aren't we working from the clone when this script is called? Need to make a copy?
+echo -en "\nCloning VDO tree $(get_repo_name ${SOURCE_REPO_URL})\n"
+git clone $SOURCE_REPO_URL ${VDO_TREE}
+
+# Loop over each commit in COMMIT_SHAS and overlay it on LINUX_SRC
+for change in ${COMMIT_SHAS[@]}; do
+  echo -en "\nProcessing changes from commit $change\n"
+
+  # Remove existing dm-vdo in case we have any deletes.
+  # XXX: The kpatch overlay process doesn't handle deletes at the moment, which
+  #      is a bug that should get handled in the future by the Makefile process
+  #      there, but it's going to get dealt with here for now.
+  if [ -d drivers/md/dm-vdo ]; then
+    rm -rf drivers/md/dm-vdo
+  fi
+
+  cd ${VDO_TREE}
+
+  # Clean up from the last build, if necessary
+  if [[ "${change}" != "${COMMIT_SHAS[0]}" ]]; then
+    echo -en "Cleaning up from the last build... "
+    git clean -fdxq
+    echo "Done"
+  fi
```

Comment:
This bit seems unnecessary given we just cloned the repo. But it might be useful in the future, so no harm leaving it for now.

Reply:
The point of this is that it runs in between processing each commit. The clone persists across all commits being processed. However, it may not be necessary since we checkout the change we intend to build. Could be that the tree is already clean due to the detached HEAD state we enter when the change is checked out.

Reply:
A new checkout won't erase build products since those aren't tracked by git (at least not the ones we've told .gitignore about, which is most of them). So the clean is probably necessary after all.